[X-Pack] Beats centralized management: security role + licensing

ycombinator opened the pull request:
Resolves #30493.

This PR adds:
* a built-in role, `beats_admin` that provides unfettered access to the `.management-beats` index. The purpose of this index is to store configuration and other peripheral information to make the Beats Centralized Management feature work.
* licensing-related logic for the Beats Centralized Management feature.

pickypg commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -217,6 +221,21 @@
+                    case STANDARD:
+                    case GOLD:
+                    case PLATINUM:
+                        return new String[] { "Logstash will no longer poll for centrally-managed configuration" };
```
Beats

pickypg commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -512,6 +531,28 @@ public boolean isLogstashAllowed() {
+     * Beats is allowed as long as there is an active license of type TRIAL, STANDARD, GOLD or PLATINUM
+     * @return {@code true} as long as there is a valid license
+     */
+    public boolean isBeatsAllowed() {
```
Given that the ingest team owns, I wonder if this should instead be `isIngestAllowed` and simply replace `isLogstashAllowed`? They share the same license requirements, so it seems reasonable to me that they shouldn't dupe their code too.

ycombinator replied:
Personally I feel like the code should reflect the products/features and not organization structure, as much as possible. So I'd like to keep this as-is.

I'd be okay with refactoring the duplicate code in the two methods into a helper method like `isNotBasic` or something like that. What do you think of that?

pickypg replied:
I'd be happy with an `isNotBasic` method.

pickypg commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/beats/BeatsFeatureSetUsage.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.io.IOException;
+
+public class BeatsFeatureSetUsage extends XPackFeatureSet.Usage {
```
It's a shame that we don't have any way for users to see if it's actually being used -- to help them to debug or even just to check it. We can add that with a follow-on PR though.

Whatever we do here should be reflected in the Logstash centralized management.

elasticmachine commented:
Pinging @elastic/es-core-infra

ycombinator commented on x-pack/plugin/core/src/main/resources/beats-admin-template.json:
```diff
@@ -0,0 +1,42 @@
+        "enrollment_token": {
+          "properties": {
+            "token": {
+              "type": "keyword"
```
Will do.

ycombinator commented on x-pack/plugin/core/src/main/resources/beats-admin-template.json:
```diff
@@ -0,0 +1,42 @@
+            }
+          }
+        },
+        "configuration_fragment": {
```
Block sounds good to me. Will change.

ycombinator commented on x-pack/plugin/core/src/main/resources/beats-admin-template.json:
```diff
@@ -0,0 +1,42 @@
+            },
+            "fragment_yml": {
+              "type": "text"
+            }
```
Sounds good, will add.

ycombinator commented on x-pack/plugin/core/src/main/resources/beats-agents-template.json:
```diff
@@ -0,0 +1,54 @@
+            "local_configuration_yml": {
+              "type": "text"
+            },
+            "central_configuration_yml": {
```
We don't have to but I think it makes sense to do it from a performance perspective. Basically we can compute this value whenever anytime changes are made in the UI, which should be relatively less frequent compared to the how often a Beat agent might request this value. So I think pre-computing it and storing it means we don't have to compute it on the fly each time a Beat requests it.

Are you thinking there is some downside of storing this? We can turn indexing off on this field if storage space is a concern (it will still be available via `_source` then).

exekias replied:
Got it, makes sense

We definitely need some kind of caching for confs, so we either do it in memory (one config per tag combination, no need to have one per beat) or store them here. I have one concern with this approach: Changing a tag requires updating all beats on that tag, that means going one by one to get their tag list, combine blocks and update the full `central_configuration_yml`.

ycombinator replied:
>  I have one concern with this approach: Changing a tag requires updating all beats on that tag, that means going one by one to get their tag list, combine blocks and update the full `central_configuration_yml`.

This is true. However, I think:
1. This should be an infrequent operation compared to how often a Beat will request its (merged) configuration,
2. We can get all the Beats for a tag from ES in a single `_search` request, and
3. We can update all the Beats' `central_configuration_yml` (after it is recomputed) in ES with a single `_bulk` request.

Let's see how all of this shakes out after an initial implement. I plan to use something like siege or httperf to simulate 10000s of Beats requesting configs while changing a tag in the midst of that scenario. We can optimize the long poles in the performance tent as we go.

ycombinator commented:
Thanks for the feedback on the schema @exekias. Yes, I imagine this schema will change as we implement the API, hence I marked this PR as WIP for now. I plan to remove the WIP and put it into formal review once all the APIs are implemented and you've had a chance to use them.

colings86 commented:
Given how similar this is to the existing Logstash ES plugin (in that it seems to just install some index templates) would it make sense to have a single ES plugin that covers both Logstash and Beats central configuration rather than having separate ones?

ycombinator commented:
@colings86 Hmmm, @pickypg also brought up something along similar lines [earlier in this PR](https://github.com/elastic/elasticsearch/pull/30520#discussion_r187383400).

I'm not opposed to sharing the _implementation_ between these two plugins in the code but I'm wary of combining their _interfaces_. Today (prior to this PR), users can disable the Logstash plugin via the `xpack.logstash.enabled: false` setting in their `elasticsearch.yml`. Also, there is a `logstash` section in the `GET http://localhost:9200/_xpack` response body. If, as part of this PR, we were to combine the `logstash` and `beats` plugins into one, those interfaces would break.

Is there some way we could keep the interfaces distinct for the `beats` and `logstash` plugins but share their implementation? If so, I can do that in this PR. But I think that might still result in having two ES plugins then, which would not accomplish your suggestion?

ycombinator commented:
Chatted with @kobelb about the security model proposed in this PR (two indices + two roles to grant different levels of access to these indices). Turns out we can simplify this model quite a bit. Specifically:

* Have just one `.management-beats` index instead of the two indices that are in the current proposal. 
* Instead of creating a new built-in `beats_agent` role, expand the built-in `kibana_system` role to have the same privileges as detailed in this proposal for the `beats_agent` role.

I will update this PR shortly to make the above changes. Thanks @kobelb!

colings86 commented:
@ycombinator I still don't really see why they need to be separate in any way. I don't think we gain much by having the ability to do `xpack.logstash.enabled:false` as opposed to telling the user that this combined plugin (lets call it `centralised-config` for now) is required for either Logstash centralised config or beats centralised config and if they don't require either Logstash or Beats centralised config they might want to do `xpack.centralised-config.enabled: false`. This is like us saying that we require `ingest-geoIP` plugin to be installed for Metricbeat, we don't have a specific plugin that the user can tell Elasticsearch that they want to use GeoIP specifically in Metricbeat, its just generally available.

I would like to understand the following:
* Both this plugin and the Logstash plugin seem to just install an index template. Is that correct?
* Are there other things that the plugin will need to do as well?
* If the plugin is just installing index templates why is it required to do this from an ES plugin rather than  something upstream installing them (i.e. whatever is using those index templates)?

The reason I ask the above is that it seems pretty heavy-weight to me to build a plugin which just effectively saves a couple of PUT index template calls and I am wondering if there is something more generic that ES could provide so its not necessary to have extra plugins here.

ycombinator commented:
@colings86 At this point I think it's safe to say these plugins only install index templates.

A bit of history here: when the Logstash plugin was originally created, there was some thinking that we might add API endpoints to it as well. Then there was talk of Kibana middleware and it was _eventually_ decided that we wouldn't be adding APIs to the Logstash plugin in ES; instead we're going to [add them in Kibana](https://github.com/elastic/kibana/issues/18185). Of course, by that time, the ES plugin was already in existence.

Now, when it came time to provide index templates for Beats central management, I followed the Logstash path for consistency and created a Beats ES plugin (via this PR). However, we know right off the bat that we won't be adding any more functionality to the Beats plugin in ES.

So, all in all, I'd be +1 to not creating a plugin for Beats in ES now. Instead we can have Kibana install the index template for Beats. I can change this PR to remove the plugin bits and leave just the bits for installing built-in Security roles.

As for the existing Logstash plugin in ES, I propose we remove it in 7.0 and have Kibana install it's index template then. This is because there could be consumers today who are using the interfaces setup by this plugin today (the ones I mentioned in https://github.com/elastic/elasticsearch/pull/30520#issuecomment-388823956 above). That way we don't introduce a breaking change in 6.x.

WDYT?

colings86 commented:
@ycombinator that sounds good to me.

ycombinator commented:
I've now removed the Beats plugin code from this PR and updated the PR title and description to accurate reflect the changeset.

ycombinator commented:
Template creation has been moved into Kibana: https://github.com/elastic/kibana/pull/19072

ycombinator commented:
@pickypg @colings86 @elastic/es-security This PR is ready for review now. Thanks!

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -205,12 +209,19 @@
-                    case GOLD:
-                    case PLATINUM:
-                        return new String[] { "Logstash will no longer poll for centrally-managed pipelines" };
+                if (!isBasic(currentMode)) {
```
do you mind switching this and other usages to `isBasic(currentMode) == false`? The elasticsearch team has a preference for this over `!`

ycombinator replied:
Fixed in d0f05d6.

jaymode commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/beats/BeatsFeatureSetUsage.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.io.IOException;
+
+public class BeatsFeatureSetUsage extends XPackFeatureSet.Usage {
```
we should make this class final

pickypg replied:
FYI, the other variants of this class are not `final`, which I think may be because some of them get mocked for tests.

ycombinator replied:
Fixed in 154d305479.

jaymode commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStoreTests.java:
```diff
@@ -220,6 +221,20 @@ public void testKibanaSystemRole() {
         });
+
+        // Beats management index
+        String index = ".management-beats";
```
can you make this variable final

ycombinator replied:
Fixed in 39d08ad742.

jaymode commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStoreTests.java:
```diff
@@ -478,6 +493,39 @@ public void testLogstashSystemRole() {
     }
 
+    public void testBeatsAdminRole() {
+        RoleDescriptor roleDescriptor = new ReservedRolesStore().roleDescriptor("beats_admin");
```
can you make this variable final

ycombinator replied:
Fixed in 39d08ad742.

jaymode commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStoreTests.java:
```diff
@@ -478,6 +493,39 @@ public void testLogstashSystemRole() {
+        assertNotNull(roleDescriptor);
+        assertThat(roleDescriptor.getMetadata(), hasEntry("_reserved", true));
+
+        Role beatsAdminRole = Role.builder(roleDescriptor, null).build();
```
can you make this variable final

ycombinator replied:
Fixed in 39d08ad742.

jaymode commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStoreTests.java:
```diff
@@ -478,6 +493,39 @@ public void testLogstashSystemRole() {
+        assertThat(beatsAdminRole.indices().allowedIndicesMatcher("indices:foo").test(randomAlphaOfLengthBetween(8, 24)),
+            is(false));
+
+        String index = ".management-beats";
```
can you make this variable final

ycombinator replied:
Fixed in 39d08ad742.

pickypg commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -205,12 +209,19 @@
+        switch (newMode) {
+            case BASIC:
+                if (!isBasic(currentMode)) {
+                    return new String[] { "Beats will no longer poll for centrally-managed configuration" };
```
This is a little misleading, which is due to how Logstash was already phrased so it may be moot. Technically Beats will continue to poll, right, but they'll see a basic license and not short circuit, then repeat later? Perhaps this should say

> Beats will no longer be able to use centrally-managed configuration

ycombinator replied:
TBH I'm not sure what the exact behavior here will be from the beats side. I bet @exekias has given this thought and can weigh in here?

exekias replied:
Technically Beats will continue polling, but won't get new configs until license is back on track. @pickypg's message may be more accurate

ycombinator replied:
Updated in d79e73d.

jaymode commented:
@ycombinator I forgot in my earlier review but there should be doc updates with this change

ycombinator commented:
> I forgot in my earlier review but there should be doc updates with this change

@jaymode Hey, sorry, but could you tell me which docs in this repo need to be updated? I was thinking of adding to the "built in role" docs but they seemed to have moved here now: https://github.com/elastic/stack-docs/blob/master/docs/en/stack/security/authorization/built-in-roles.asciidoc. I am working on a PR to update those. Are there any other docs, in this (elastic/elasticsearch) repo that I should be updating too? Thanks.

ycombinator commented:
@jaymode The PR I mentioned in my previous comment is up now: https://github.com/elastic/stack-docs/pull/80.

ycombinator commented:
We'll be backporting this to `6.x` whenever the rest of the Beats management feature is ready (or close to being ready). I'm tracking this pending backport in the Beats management meta issue: https://github.com/elastic/beats/issues/7028

jaymode commented:
@ycombinator is there an expected version? I added the v7 label to this since it went into master. Usually we also add the 6.x version label too that it is supposed to go in. Since it is not backported yet, you would also add the backport pending label and remove that label once the backport is complete

ycombinator commented:
> is there an expected version? ... Usually we also add the 6.x version label too that it is supposed to go in.

I believe we're targeting 6.5, so I've added that label now.

> Since it is not backported yet, you would also add the backport pending label and remove that label once the backport is complete

Yep, added.
